source/examples/cppfs-watch/main.cpp

```diff
@@ -0,0 +1,132 @@
+#include <cppfs/LoginCredentials.h>
+#include <cppfs/cppfs-version.h>
+#include <cppfs/fs.h>
+
+
+using namespace cppassist;
+using namespace cppfs;
+
+
+int main(int argc, char * argv[])
+{
+    // Declare program
+    CommandLineProgram program(
+        "cppfs-watch",
+        "cppfs-watch " CPPFS_VERSION,
+        "Listen to events in the file system and print them to the console."
+    );
+
+    CommandLineAction action("watch", "Watch files in directory");
+    program.add(&action);
+
+    CommandLineSwitch swHelp("--help", "-h", "Print help text", CommandLineSwitch::Optional);
+    action.add(&swHelp);
+
+    CommandLineOption opConfig("--config", "-c", "file", "Load configuration from file", CommandLineOption::Optional);
+    action.add(&opConfig);
+
+    CommandLineOption opTime("--time", "-t", "seconds", "Stop watch loop after given time", CommandLineOption::Optional);
+    action.add(&opTime);
+    
+
+    CommandLineParameter paramPath("path", CommandLineParameter::Optional);
+    action.add(&paramPath);
+
+    // Parse command line
+    program.parse(argc, argv);
+    if (program.hasErrors() || swHelp.activated())
+    {
+        // Print help and exit
+        program.print(program.help());
+        return 0;
+    }
+
+    try {
+        // Get login credentials
+        LoginCredentials login;
+
+        std::string configFile = opConfig.value();
+        if (!configFile.empty())
+        {
+            login.load(configFile);
+        }
+
+        // Get path
+        std::string path = paramPath.value();
+        if (path.empty()) path = ".";
+
+        // Open directory
+        FileHandle dir = fs::open(path, &login);
+        if (dir.isDirectory())
+        {
+            // Watch directory
+            FileWatcher watcher = dir.watch();
+
+            // Create file event handler
+            watcher.addHandler([] (FileHandle & fh, FileEvent event) {
+                // Get file type
+                std::string type = (fh.isDirectory() ? "directory" : "file");
+
+                // Get operation
+                std::string operation = ( (event & FileCreated) ? "created" :
+                                        ( (event & FileRemoved) ? "removed" :
+                                                                  "modified" ) );
+
+                // Log event
+                std::cout << "The " << type << " '" << fh.path() << "' was " << operation << "." << std::endl;
+            });
+
+            // Begin watching and printing events
+            std::string t = opTime.value();
+            if (t.empty()) {
+                while (true) {
+                    watcher.watch();
+                }
+            }
+            else {
+                auto t_sleep = std::stoi(t);
+                std::cout << "Will stop watching after " << t_sleep << " seconds..." << std::endl;
+
+                // Execute watch loop in a separate thread
+                std::atomic<bool> stop_thread{false};
```

Comment:
Why use a thread? We have the timeout, so just have the loop around it and stop after n rounds. This is not as precise, but multithreading is also never precise. I would always avoid multithreading when it's not necessary.

Reply:
Because in my case, this would always be the case, i.e. having a separate thread that does the polling of the watcher object, so it put it there mainly for testing purposes. Remove at your discretion.